Improve node failure

Vicente-Cheng opened the pull request:
<!-- 
!IMPORTANT!
Please do not create a Pull Request without creating an issue first.
-->

#### Problem:
Improve our mechanism when node failed

#### Solution:
Improve our mechanism when node failed

#### Related Issue(s):
https://github.com/harvester/harvester/issues/8945

#### Test plan:
1. create 3 node cluster
2. create VM
3. disable the network on the host which the VM is running
4. ensure the VM can be migrated smoothly
5. ensure node which network was disabled have the following taints
`kubevirt.io/drain:NoSchedule`
`node.kubernetes.io/out-of-service`
6. enable the network. Ensure the above taints were removed after the node was back.

#### Additional documentation or context

Vicente-Cheng committed:
controller: improve the host failure case

    Now we will handle the Kubelet/Node failure cases.
    We will leverage the kubevirt taint to trigger the migration and
    out-of-service to clean up the resource.

    That would help speed up the migration process when a node fails.

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

brandboat commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,315 @@ package node
-	if !vmForceResetPolicy.Enable {
-		return node, nil
+// just logger when we detect node is under disk pressure, maybe we can do more things in the future
+func (h *nodeDownHandler) check_disk_pressure(node *corev1.Node) error {
```
nit:
```suggestion
func (h *nodeDownHandler) checkDiskPressure(node *corev1.Node) error {
```

brandboat commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,315 @@ package node
+
+// get_ip_from_other_nodes tries to get an reachable IP from other nodes
+// we directly get this annotation `etcd.rke2.cattle.io/node-address` set by RKE2
+func get_ip_from_other_nodes(currentNode *corev1.Node, nodes []*corev1.Node) string {
```
nit:
```suggestion
func getIpFromOtherNodes(currentNode *corev1.Node, nodes []*corev1.Node) string {
```

Vicente-Cheng replied:
updated

brandboat commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,315 @@ package node
+	pinger, err := ping.NewPinger(ip)
+	if err != nil {
+		// create Pinger failed, skip testing, assume node is reachable
+		return connected, fmt.Errorf("create pinger failed, error: %s", err.Error())
```
nit:
```suggestion
		return connected, fmt.Errorf("create pinger failed, error: %w", err)
```

ibrokethecloud commented on pkg/settings/settings_helper.go:
```diff
@@ -67,7 +67,7 @@ func (target *BackupTarget) IsDefaultBackupTarget() bool {
 	policy := &VMForceResetPolicy{
 		Enable: true,
-		Period: 5 * 60, // 5 minutes
+		Period: 30, // 30 seconds
```
the default originally was 300 seconds. just wanted to make sure if this is intentional?

Vicente-Cheng commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,315 @@ package node
+			if err := h.removeTaints(node, taintKeys...); err != nil {
+				return fmt.Errorf("failed to remove taints %v from node %s: %w", taintKeys, node.Name, err)
+			}
+		}
```
updated, thanks for the suggestion!

Vicente-Cheng commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,315 @@ package node
+	if cond != nil && time.Since(cond.LastTransitionTime.Time) < targetInterival {
+		deadline := cond.LastTransitionTime.Add(targetInterival)
+		logrus.Debugf("Enqueue node event again at %v", deadline)
+		h.nodes.EnqueueAfter(node.Name, time.Until(deadline))
```
nice catch!
updated

Vicente-Cheng commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,315 @@ package node
-	if !vmForceResetPolicy.Enable {
-		return setting, nil
+	if vmForceResetPolicy.Enable {
+		targetInterival = time.Duration(vmForceResetPolicy.Period) * time.Second
```
I prefer to keep that default on the nodeDown controller once we change the default value for `vmForceResetPolicy`.
WDYT?

Vicente-Cheng replied:
No, I mean keep the defaultTimeInterval on installer if user set `vmForceResetPolicy.Enabled` to false, we can still have the default timeout value here.

brandboat replied:
Got it, thanks for the explanation. I'm fine with the change.

Vicente-Cheng commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,315 @@ package node
-	vmForceResetPolicy, err := settings.DecodeVMForceResetPolicy(setting.Value)
+func (h *nodeDownHandler) tryWaitForKubelet(node *corev1.Node) (bool, error) {
+	// get VMForceResetPolicy setting
+	targetInterival := defaultTimeInterval
```
udpated, thanks!

Vicente-Cheng commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,315 @@ package node
+			logrus.Warnf("can't get reachable IP from other nodes for node %s, skip this time", node.Name)
+			return nil
+		}
+		connected, err := pingNode(targetIP)
```
we drop piner

Vicente-Cheng commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,315 @@ package node
+// here we need to check node ready condition, two cases here:
+// 1. node not reachable -> we can handle it directly
+// 2. kubelet not ready -> we need to wait for vmForceResetPolicy.Period seconds before handling
+func (h *nodeDownHandler) check_node_ready(node *corev1.Node) error {
```
updated

brandboat reviewed (dismissed):
LGTM, thanks!

brandboat commented:
gentle ping @Vicente-Cheng, now we dropped the go-ping lib, could you please remove the related libs under vendor? Thanks. ([e0f3732] not required anymore)

WebberHuang1118 reviewed (dismissed):
LGTM, just a question, thanks.

ihcsim commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,239 @@ package node
 const (
 	nodeDownControllerName = "node-down-controller"
+	kubevirtDrainTaintKey  = "kubevirt.io/drain"
+	nonGracefulTaintKey    = "node.kubernetes.io/out-of-service"
```
can we use the `corev1.TaintNodeOutOfService` as the value, instead of the literal?

ihcsim replied:
Consider using the upstream consts instead of literals:

* `kubevirt.io/kubevirt/pkg/virt-config.NodeDrainTaintDefaultKey`
* `corev1.TaintNodeOutOfService`

ihcsim commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,239 @@ package node
-	}
+		// we wait enough or node is unreachable, add taints to trigger migration
+		if getNodeTaint(node.Spec.Taints, kubevirtDrainTaintKey) == nil {
+			if err := h.triggerMigration(node); err != nil {
```
I think it's not necessary to trigger migration because the VMs are configured with `.spec.evictionStrategy: LiveMigrateIfPossible`. From my brief testing, the kubevirt auto-starts the live migration the moment when the node is marked as out-of-service:

* create 2-node cluster
* start a vm on the 2nd non-primary node
* stop kubelet on the node where vm is running: `sudo systemctl stop rke2-agent`
* watch `kubectl get node` output and see 2nd node transitions to `NotReady` state
* add out-of-service taint to 2nd node - live migration happens immediately

brandboat replied:
I have the same observation in my local env.

Vicente-Cheng replied:
I do some tests on it, if we directly taint the out-of-service, the pod would be delete directly instead of migration.
So we still need kubevirt drain taint.

And I thought your case might be related to the kubevirt heartbeat, once kubevirt heartbeat timeout, it also trigger migration.

brandboat replied:
Thanks for the explanation. Just to confirm—if a node goes down, we’ll have to wait until the migration timeout expires before the VM can come back online, correct?

Vicente-Cheng committed:
settings: add VMMigrationTimeout fields

    - add VMMigrationTimeout to VMForceResetPolicy

    We need this fields to let user configure how much extra time they
    need to wait before add out-of-service taint.

Signed-off-by: Vicente Cheng <vicente.cheng@suse.com>

Vicente-Cheng commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,239 @@ package node
-	cond := getNodeCondition(node.Status.Conditions, corev1.NodeReady)
-	if cond == nil {
-		return node, fmt.Errorf("can't find %s condition in node %s", corev1.NodeReady, node.Name)
+	if err := h.checkDiskPressure(node); err != nil {
```
Thanks, I added the comment and drop the diskpressure operation

Vicente-Cheng commented:
@mergifyio backport v1.7

ihcsim commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,379 @@ package node
+	targetInterval := time.Duration(vmMigrationTimeout) * time.Second
+	if cond == nil {
+		// means we need to change the condition to unreachable (add a new one or update the existing one)
+		if err := h.setHarvesterNodeDrainedCondition(node, corev1.ConditionFalse, HarvesterNodeFailureCondReasonUnreachable, HarvesterNodeFailureCondMsgUnreachable); err != nil {
```
Re: the `reason` for this condition (here and in line 131), normally, we specify the reason (not the cause) for the true/false status of a condition. In our case, the node drained condition is false because draining is still in progress or the timeout hasn't expired. Conversely in line 131, the node drained condition is true because draining is completed or the timeout has expired. 

Node failure is the cause that triggers this condition (which can be the `message`), not the reason why the condition is in a certain state.

ibrokethecloud commented on pkg/settings/settings_helper.go:
```diff
@@ -78,8 +81,19 @@ func InitVMForceResetPolicy() string {
+	type oldPolicy struct {
+		Enable bool `json:"enable"`
+		Period int  `json:"period"`
+	}
```
we can make the `VMMigrationTimeout` as optional field, and then same object can be used to unmarshal new or old settings, and we can apply default if one does not exist

Vicente-Cheng replied:
thanks for the tips, updated

ibrokethecloud reviewed (commented):
the two methods `tryWaitForExtraTimeout` and `tryWaitForVMMigration` have a lot of common logic which could be factored into a common method say `waitForConditionTimeout(node *corev1.Node, conditionType corev1.NodeConditionType, timeout int64)` and all this method does is requeue and return if condition has not expired.

This way object keeps getting requeued until timeout has expired else code flow continues, and will make it easier to track how the node object is being requeued to ensure we are waiting for timeout to occur.

we should also call the method `setHarvesterNodeDrainedCondition`  as soon as kubevirt drain taint is called. 

this will ensure `addOutOfServiceTaintAfterMigrationTimeout` is only looking for timeout of our custom node drain condition before marking node out of service.

brandboat commented on pkg/controller/master/node/node_down_controller.go:
```diff
@@ -3,278 +3,343 @@ package node
+			return nil
+		}
+
+		node, err = h.createOrUpdateNodeCondition(node, HarvesterNodeCondDrained, corev1.ConditionFalse, HarvesterNodeDrainedCondReasonDraining, HarvesterNodeDrainedCondMsg)
```
Will this cause busy loop when node down?  There is no reflect.deepEqual check in `createOrUpdateNodeCondition`, when we invoke `createOrUpdateNodeCondition` it will trigger another reconcile loop in this controller, right?

brandboat reviewed (approved):
LGTM, thanks!

Vicente-Cheng commented:
@mergifyio backport v1.7
